### ISSUE TITLE
[Question] Using `DQN`/`PPO` to run `Pong`: No good results.

### ISSUE THREAD

**Issue by LainWeigh** [DLR-RM/stable-baselines3#1446]
### ❓ Question

Hi, all.

I'm running `Pong` in `Atari` with the `DQN`/`PPO` algorithm with the parameters provided in `rl-baselines3-zoo`. However, I failed to run it, and the result shows that the agent only outputs one action (such as moving up) at all times.

I suspect it is because the features extracted by the `CNN` network are wrong. I checked the output of the `CNN` network at each timestep and found that they are almost the same.

why is that? How would I improve the `CNN` network structure, or other structures of the network?

Thanks.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by qgallouedec** [comment]
Can you share your results and your hyperparms? (at least those who are different from rl-zoo3).

With the hyperparams of rl-zoo3, here are the results (see https://wandb.ai/openrlbenchmark/sb3):

![W B Chart 13_04_2023, 11_38_16-2](https://user-images.githubusercontent.com/45557362/231720033-6376171c-7bbb-48cf-9d63-a0d372b39ad2.png)

**Comment by LainWeigh** [comment]
Thank you for such a quick reply!

Sorry I didn't document the curves, but I can describe how the results change in each rollout. In each rollout, the value of 'ep_rew_mean' is around -20 (such as -20.3, -20.4, -20.6, etc.), which seems to indicate that the agent has not learned the policy at all.

**Comment by qgallouedec** [comment]
So you don't use rl-zoo3 then?
Your question is also about A2C? What env do you use? Please provide the full code.

**Comment by araffin** [comment]
If you want to reproduce the result from the RL benchmark:
```
python3 -m rl_zoo3.train --algo ppo --env PongNoFrameskip-v4 -f logs/
```

See https://huggingface.co/sb3/ppo-PongNoFrameskip-v4
and https://huggingface.co/sb3/a2c-PongNoFrameskip-v4
or https://wandb.ai/openrlbenchmark/sb3/runs/zs4hn2hs?workspace=

**Comment by LainWeigh** [comment]
Thanks!

Sorry, I've tried `A2C`, `PPO` and `DQN`, but no one has good result. Taking A2C as an example, I use the parameters in the "Hyperparameters" section of the URL [](https://huggingface.co/sb3/a2c-PongNoFrameskip-v4):
```
OrderedDict([('ent_coef', 0.01),
             ('env_wrapper',
              ['stable_baselines3.common.atari_wrappers.AtariWrapper']),
             ('frame_stack', 4),
             ('n_envs', 16),
             ('n_timesteps', 10000000.0),
             ('policy', 'CnnPolicy'),
             ('policy_kwargs',
              'dict(optimizer_class=RMSpropTFLike, '
              'optimizer_kwargs=dict(eps=1e-5))'),
             ('vf_coef', 0.25),
             ('normalize', False)])
```

But I did not use the terminal to run, nor did I load the pre-trained model, I used the following python code:

```
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.sb2_compat.rmsprop_tf_like import RMSpropTFLike
from stable_baselines3.common.vec_env import VecFrameStack
from stable_baselines3 import A2C

# There already exists an environment generator
# that will make and wrap atari environments correctly.
# Here we are also multi-worker training (n_envs=4 => 4 environments)
env = make_atari_env("PongNoFrameskip-v4", n_envs=16, seed=0)
# Frame-stacking with 4 frames
env = VecFrameStack(env, n_stack=4)
model = A2C("CnnPolicy", env, verbose=1, ent_coef=0.01,
            policy_kwargs=dict(optimizer_class=RMSpropTFLike, optimizer_kwargs=dict(eps=1e-5)),
            vf_coef=0.25)


model.learn(total_timesteps=100000)
model.save("dqn_atari2")
```

But the result is that the value of reward has not increased, and it has always remained at around -20.

**Comment by araffin** [comment]
RL Zoo:
n_timesteps', 10000000.0 -> 10^7

Your script :

total_timesteps=100000  -> 10^5

Your budget is 100x smaller than the recommended one...

**Comment by LainWeigh** [comment]
I've tried 10^7 trials before this, but the result is still the same(around -20), and the reward didn't grow.

**Comment by LainWeigh** [comment]
And this line chart shows that the reward always grows in the first 2M timesteps, but my result is not like this.
![image](https://user-images.githubusercontent.com/126131200/231953838-1fc3b4a1-444f-415e-b6a8-d81cb24efb69.png)

**Comment by araffin** [comment]
> If you want to reproduce the result from the RL benchmark:
> 
> ```
> python3 -m rl_zoo3.train --algo ppo --env PongNoFrameskip-v4 -f logs/
> ```
> 
> See https://huggingface.co/sb3/ppo-PongNoFrameskip-v4 and https://huggingface.co/sb3/a2c-PongNoFrameskip-v4 or https://wandb.ai/openrlbenchmark/sb3/runs/zs4hn2hs?workspace=

I just tried the command with latest SB3 version (1.8.0), and I could get working runs for both A2C and PPO:
https://wandb.ai/araffin/sb3/runs/di7bd5mp?workspace=user-araffin
and
https://wandb.ai/araffin/sb3/runs/et7kh6gv?workspace=user-araffin

I will try with your script.

**Comment by araffin** [comment]
I also tested with that script and it works in my side:

```python
from rl_zoo3.utils import linear_schedule
from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.vec_env import SubprocVecEnv, VecFrameStack


if __name__ == "__main__":
    env_id = "PongNoFrameskip-v4"
    env = make_atari_env(env_id, n_envs=8, seed=0, vec_env_cls=SubprocVecEnv)
    env = VecFrameStack(env, n_stack=4)

    hyperparams = {
        "n_steps": 128,
        "ent_coef": 0.01,
        "learning_rate": linear_schedule(2.5e-4),
        "n_epochs": 4,
        "clip_range": linear_schedule(0.1),
        "vf_coef": 0.5,
        "tensorboard_log": "/tmp/sb3/",
        "policy_kwargs": None,
        "verbose": 1,
    }

    # Use tuned hyperparameters from the RL Zoo
    model = PPO("CnnPolicy", env, **hyperparams)

    model.learn(total_timesteps=int(1e7), progress_bar=True, log_interval=50)
```

**Comment by LainWeigh** [comment]
Thanks so much! As you said, I increased the timesteps and the results kept getting better and the reward value kept increasing, whether `PPO` or `A2C`. It seems that the steps required by `CNNPolicy` will be much longer.

Thank you for your patient reply!
